kernel/sched.c

```diff
@@ -1347,3 +1378,78 @@ int k_thread_cpu_mask_disable(k_tid_t thread, int cpu)
 }
 
 #endif /* CONFIG_SCHED_CPU_MASK */
+
+int z_impl_k_thread_join(struct k_thread *thread, s32_t timeout)
+{
+	k_spinlock_key_t key;
+	int ret;
+
+	__ASSERT(((arch_is_in_isr() == false) || (timeout == K_NO_WAIT)), "");
+
+	key = k_spin_lock(&sched_spinlock);
+
+	if ((thread->base.thread_state & _THREAD_DEAD) != 0) {
```

Comment:
This looks racy.  What if a thread terminates after this line but before you add yourself to its queue?

Reply:
Oh wait, no, you're locked here.  This one is fine.